lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -58,6 +58,15 @@ class TensorEmptyConversionPattern
   matchAndRewrite(tensor::EmptyOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
 
+    // BroadcastOp will be folded away so its EmptyOp should be erased.
+    if (not op->getUsers().empty()) {
+      Operation *emptyOp = *op->getUsers().begin();
```

Comment:
It's a bit confusing to name this variable `emptyOp` shouldn't it be `broadcastOp`?  We also only check the first user, don't we need to check if all users are broadcasts?

Also, I think we should rely on vanilla MLIR pass to clean up the straggling empties.  Can't we rely on that?

Reply:
I just fixed the variable name, please refresh. Let me update to check for all users. 

Yes I was hoping that we run remove-dead-values but I see that we don't run it at the moment. I tried to enabled it for TTIRToTNN pipeline but it didn't resolve this issue. Then I found the following comment from @mrakitaTT  so I chose to explicit remove the EmptyOp.

```
Option<bool> removeDeadValuesEnabled{
*this, "enable-remove-dead-values",
llvm::cl::desc("Enable --remove-dead-values optimization pass."),
// Currently this pass fails if module has a name, so keeping the
// optimization OFF by default until that issue is fixed on llvm side.
```

Reply:
@mrakitaTT, are we able to run this pass?

Reply:
So I just tried adding the remove-dead-values pass again and I see that although the pass does get executed, but it doesn't remove the dead instruction:

`%3 = "ttnn.empty"(%0) <{dtype = #tt.supportedDataTypes<f32>, layout = #ttnn.layout<tile>, memory_config = #ttnn.memory_config<<interleaved>, <dram>>, shape = #ttnn.shape<512x512>}> : (!tt.device<#device>) -> tensor<512x512xf32, #layout3>`

IR log:

```
// -----// IR Dump After RemoveDeadValues (remove-dead-values) //----- //
#device = #tt.device<workerGrid = #tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, l1Map = (d0, d1)[s0, s1] -> (0, d0 floordiv s0, d1 floordiv s1, (d0 mod s0) * s1 + d1 mod s1), dramMap = (d0, d1)[s0, s1] -> (0, 0, ((((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 8192) mod 12, (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) floordiv 98304 + (((d0 floordiv s0) * 8 + d1 floordiv s1) * (s1 * s0) + (d0 mod s0) * s1 + d1 mod s1) mod 8192), meshShape = , chipIds = [0]>
#dram = #tt.memory_space<dram>
#system = #tt.memory_space<system>
#system_desc = #tt.system_desc<[{arch = <wormhole_b0>, grid = 8x8, l1_size = 1499136, num_dram_channels = 12, dram_channel_size = 1073741824, noc_l1_address_align_bytes = 16, pcie_address_align_bytes = 32, noc_dram_address_align_bytes = 32, l1_unreserved_base = 1024, erisc_l1_unreserved_base = 1024, dram_unreserved_base = 1024, dram_unreserved_end = 1073741824, physical_cores = {worker = [ 0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  1x0,  1x1,  1x2,  1x3,  1x4,  1x5,  1x6,  1x7,  2x0,  2x1,  2x2,  2x3,  2x4,  2x5,  2x6,  2x7,  3x0,  3x1,  3x2,  3x3,  3x4,  3x5,  3x6,  3x7,  4x0,  4x1,  4x2,  4x3,  4x4,  4x5,  4x6,  4x7,  5x0,  5x1,  5x2,  5x3,  5x4,  5x5,  5x6,  5x7,  6x0,  6x1,  6x2,  6x3,  6x4,  6x5,  6x6,  6x7,  7x0,  7x1,  7x2,  7x3,  7x4,  7x5,  7x6,  7x7] dram = [ 8x0,  9x0,  10x0,  8x1,  9x1,  10x1,  8x2,  9x2,  10x2,  8x3,  9x3,  10x3]}, supported_data_types = [<f32>, <f16>, <bf16>, <bfp_f8>, <bfp_bf8>, <bfp_f4>, <bfp_bf4>, <bfp_f2>, <bfp_bf2>, <u32>, <u16>, <u8>], supported_tile_sizes = [ 4x16,  16x16,  32x16,  4x32,  16x32,  32x32]}], [0], [3 : i32], [ 0x0x0x0]>
#layout = #tt.layout<(d0) -> (0, d0), undef, <1x1>, memref<1x1xf32, #system>>#layout1 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<512x512xf32, #system>>#layout2 = #tt.layout<(d0) -> (0, d0), undef, <1x1>, memref<1x1x!tt.tile<32x32, f32>, #dram>, interleaved>#layout3 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<16x16x!tt.tile<32x32, f32>, #dram>, interleaved>
module @jit_broadcast attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32, tt.device = #device, tt.system_desc = #system_desc} {  func.func public @main(%arg0: tensor<1xf32, #layout> {mhlo.layout_mode = "default"}, %arg1: tensor<512x512xf32, #layout1> {mhlo.layout_mode = "default"}) -> (tensor<512x512xf32, #layout1> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = "ttnn.get_device"() <{mesh_shape = #ttnn<mesh_shape 1x1>}> : () -> !tt.device<#device>
    %1 = "ttnn.to_layout"(%arg0, %0) <{layout = #ttnn.layout<tile>}> : (tensor<1xf32, #layout>, !tt.device<#device>) -> tensor<1xf32, #layout2>
    %2 = "ttnn.to_device"(%1, %0) <{memory_config = #ttnn.memory_config<<interleaved>, <dram>>}> : (tensor<1xf32, #layout2>, !tt.device<#device>) -> tensor<1xf32, #layout2>
    %3 = "ttnn.empty"(%0) <{dtype = #tt.supportedDataTypes<f32>, layout = #ttnn.layout<tile>, memory_config = #ttnn.memory_config<<interleaved>, <dram>>, shape = #ttnn.shape<512x512>}> : (!tt.device<#device>) -> tensor<512x512xf32, #layout3>
    %4 = "ttnn.to_layout"(%arg1, %0) <{layout = #ttnn.layout<tile>}> : (tensor<512x512xf32, #layout1>, !tt.device<#device>) -> tensor<512x512xf32, #layout3>
    %5 = "ttnn.to_device"(%4, %0) <{memory_config = #ttnn.memory_config<<interleaved>, <dram>>}> : (tensor<512x512xf32, #layout3>, !tt.device<#device>) -> tensor<512x512xf32, #layout3>
    %6 = "ttnn.empty"(%0) <{dtype = #tt.supportedDataTypes<f32>, layout = #ttnn.layout<tile>, memory_config = #ttnn.memory_config<<interleaved>, <dram>>, shape = #ttnn.shape<512x512>}> : (!tt.device<#device>) -> tensor<512x512xf32, #layout3>
    %7 = "ttnn.maximum"(%2, %5, %6) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<1xf32, #layout2>, tensor<512x512xf32, #layout3>, tensor<512x512xf32, #layout3>) -> tensor<512x512xf32, #layout3>
    %8 = "ttnn.to_memory_config"(%7, %0) : (tensor<512x512xf32, #layout3>, !tt.device<#device>) -> tensor<512x512xf32, #layout1>
    return %8 : tensor<512x512xf32, #layout1>
  }
}
```

Reply:
I spent some time looking at this further by debugging the remove-dead-values pass, it seems it exits early due to failing this function:

```
bool mlir::isMemoryEffectFree(Operation *op) {
  if (auto memInterface = dyn_cast<MemoryEffectOpInterface>(op)) {
    if (!memInterface.hasNoEffect())
      return false;
    // If the op does not have recursive side effects, then it is memory effect
    // free.
    if (!op->hasTrait<OpTrait::HasRecursiveMemoryEffects>())
      return true;
  } else if (!op->hasTrait<OpTrait::HasRecursiveMemoryEffects>()) {
    // Otherwise, if the op does not implement the memory effect interface and
    // it does not have recursive side effects, then it cannot be known that the
    // op is moveable.
    return false;
  }
```

I tried removing this check followed by several others and it still keeps asserting for several reasons in the Verifier and AsmPrinter. It makes me wonder if we ever used any MLIR vanilla pass before? We need some work to enable MLIR passes, maybe we should add that as a TODO item otherwise we are loosing on MLIR goodies.

Reply:
The empty op is marked with trait:
```
// Always speculatable operation that does not touch memory.  These operations                                                  
// are always legal to hoist or sink.                                                                             
def Pure : TraitList<[AlwaysSpeculatable, NoMemoryEffect]>;
```

This should make it legal to erase automatically by MLIR infra.  It seems like @mrakitaTT disabled remove dead values for some reason.

Let's wait for him to return before making further progress on this PR.  I'd rather have the robust solution of MLIR removing dead values vs us manually cleaning up edge cases like this.

Reply:
It seems that `DCE` has bug which exposes if we have named modules

Marko reported it with: https://github.com/llvm/llvm-project/issues/107870
It seems that its fixed: https://github.com/llvm/llvm-project/pull/109990

We should uplift llvm version...

Reply:
Yeah let's uplift LLVM

Reply:
I didn't see the module name error because I was already using an uplifted LLVM version required for torch-mlir build, opened a PR for the same.